### DISCUSSION TITLE
Why does CH need so many connections?

### DISCUSSION THREAD

**Discussion by IronNail** [ClickHouse/ClickHouse#60222] [Q&A]
Hello!

ClickHouse version = 23.10.1.1976

**Description:**
I have created a dictionary which gets data from SQL Server:
`CREATE DICTIONARY db.Calendars
(
CalendarID UUID,BusinessDate DateTime,WeekStart DateTime,WeekEnd DateTime
) PRIMARY KEY CalendarID, BusinessDate
SOURCE(ODBC(
CONNECTION_STRING 'DSN=<serverName>;UID=RMDCCHReader...;PWD={<password>};Schema=dbo;Database=db;Timeout=600;' TABLE 'DWCalendar'
)) LIFETIME(MIN 0 MAX 0)
LAYOUT(COMPLEX_KEY_HASHED())`

Then I checked, that there are no active connections to SQL Server from CH.
Then I fill the dictionary with command 
`SYSTEM RELOAD DICTIONARY db.Calendars`

CH creates 3-4 connections which are calling the same query
![image](https://github.com/ClickHouse/ClickHouse/assets/145397088/f7032cda-3708-459b-9dd5-b25abcdd9c95)

When CH finishes, all 3-4 connections stay at status "sleeping" for a long time.
When I call dictionary reloading once again, I see that CH reuses them but not all of them. And again, it runs the same command in parallel.

**Questions**
1) Does CH really need 3-4 connections at the same time?
2) Why do I see the same running command for all active sessions?
3) Why does CH not break connections, when it finishes data loading process?
4) Is it planned behavior? :)

**Comment by den-crane** [comment]
It uses a connection pool with permanent connections to reduce a number session create events which are expensive, see connection pool settings.

see `last_since` column in your picture, the query was executed at a different time, using different connections.

**Reply by IronNail** [reply]
Ok, I understand about connections life-time. It has a sense.
But why do I see the same query in different ACTIVE sessions? It means that CH runs a quite expensive query 3 times in parallel.

**Reply by den-crane** [reply]
Not sure what you mean active? CH session? system.processes ?

**Reply by IronNail** [reply]
Please look at my screen shoot. There are three sessions on SQL Server, which were created by CH, and all of them are active at the same time. The second table shows what they are doing (input_buffer, current_sp columns). They all are running the same query which provides data for dictionary. Why does CH need to run the same query 3 times in parallel on the SQL Server?
As for me, it looks strange.

**Reply by den-crane** [reply]
It could be because the query is executed too long, and CH retries the query

**Reply by IronNail** [reply]
Great! This query really runs for 3-9 minutes depending on SQL Server loading. 
Thanks for your answers, time and patience.

**Reply by den-crane** [reply]
Then it's expected. As I remember there is a tunable timeout, by default a dictionary waits for 60 seconds.

**Reply by IronNail** [reply]
Do you know how to change this setting? I want to increase it.
